sshd-core/src/main/java/org/apache/sshd/client/channel/ChannelX11.java

```diff
@@ -0,0 +1,166 @@
+
+    private final AtomicBoolean isInitialized = new AtomicBoolean(false);
+    private final String host;
+    private final int port;
+
+    private IoSession x11;
+
+    public ChannelX11(String host, int port) {
+        super("x11");
+        this.host = host;
+        this.port = port;
+    }
+
+    @Override
+    public OpenFuture open(long recipient, long rwSize, long packetSize, Buffer buffer) {
+        final DefaultOpenFuture defaultOpenFuture = new DefaultOpenFuture(this, futureLock);
+        super.openFuture = defaultOpenFuture;
+
+        final IoConnector connector
+                = getSession().getFactoryManager().getIoServiceFactory().createConnector(createX11IoHandler());
+        addCloseFutureListener(future -> {
+            if (future.isClosed()) {
+                connector.close(true);
+            }
+        });
+
+        final IoConnectFuture connectFuture = connector.connect(new InetSocketAddress(host, port), this, null);
+        connectFuture.addListener(future -> {
+            if (future.isConnected()) {
+                x11 = future.getSession();
+                handleOpenSuccess(recipient, rwSize, packetSize, buffer);
+            } else {
+                if (future.getException() != null) {
+                    defaultOpenFuture.setException(future.getException());
+                } else {
+                    defaultOpenFuture.setValue(false);
+                }
+                unregisterSelf();
+            }
+        });
+
+        return defaultOpenFuture;
+
+    }
+
+    @Override
+    protected void doOpen() throws IOException {
+        setOut(new ChannelOutputStream(this, getRemoteWindow(), log, SshConstants.SSH_MSG_CHANNEL_DATA, true));
+    }
+
+    @Override
+    protected void doWriteData(byte[] data, int off, long len) throws IOException {
+        if (isInitialized.compareAndSet(false, true)) {
+            final byte[] xCookie = getXCookie();
+            if (xCookie == null) {
+                sendEof();
+                return;
+            }
+
+            int l = (int) len;
+            int s = 0;
+            byte[] foo = new byte[l];
+            System.arraycopy(data, off, foo, 0, l);
+
+            if (l < 9) {
+                return;
+            }
+
+            int plen = (foo[s + 6] & 0xff) * 256 + (foo[s + 7] & 0xff);
+            int dlen = (foo[s + 8] & 0xff) * 256 + (foo[s + 9] & 0xff);
+
+            if ((foo[s] & 0xff) == 0x6c) {
+                plen = ((plen >>> 8) & 0xff) | ((plen << 8) & 0xff00);
+                dlen = ((dlen >>> 8) & 0xff) | ((dlen << 8) & 0xff00);
+            }
+
+            if (l < 12 + plen + ((-plen) & 3) + dlen) {
+                return;
+            }
+
+            byte[] bar = new byte[dlen];
+            System.arraycopy(foo, s + 12 + plen + ((-plen) & 3), bar, 0, dlen);
+
+            if (Objects.deepEquals(xCookie, bar)) {
+                x11.writeBuffer(new ByteArrayBuffer(foo, s, l));
```

Comment:
If the cookie is the one that was random-generated below, then you would have to replace it here with the real cookie. So there must be a way to set that real cookie on the channel. Client code could read it from the local .Xauthority file.